apps/state/lib/state/vehicle.ex

```diff
@@ -66,12 +66,41 @@ defmodule State.Vehicle do
     State.Vehicle.select(matchers, idx)
   end
 
-  defp get_index(%{labels: labels}) when labels != [], do: :label
   defp get_index(%{routes: routes}) when routes != [], do: :effective_route_id
   defp get_index(_filters), do: nil
 
   defp build_filters(matchers, _key, nil, _filters), do: matchers
 
+  defp build_filters(matchers, :label, labels, _filters) do
+    consist_matches? = fn %Model.Vehicle{consist: consist}, labels ->
+      case consist do
+        nil ->
+          false
+
+        _ ->
+          labels
+          |> MapSet.new()
+          |> MapSet.intersection(consist)
+          |> Kernel.!==(MapSet.new())
+      end
+    end
+
+    label_matches? = fn %Model.Vehicle{label: label}, labels ->
+      label in labels
+    end
+
+    vehicle_ids =
+      all()
+      |> Enum.filter(fn vehicle ->
+        label_matches?.(vehicle, labels) or consist_matches?.(vehicle, labels)
+      end)
+      |> Enum.map(& &1.id)
+
+    for matcher <- matchers, vehicle_id <- vehicle_ids do
+      Map.put(matcher, :id, vehicle_id)
+    end
+  end
```

Comment:
If you want to go this route, it would be better to do this as a post-filter. What you're doing here is querying all the vehicles, filtering them, then using the IDs to query the vehicles. But you already had the vehicles after the second step, so the extra ID filter is wasted work.

Other options would be to do either something like what we do with Alerts (separate tables for querying parts of the data) or Trips (extra fields to handle the fact that a vehicle can have multiple labels.